[Feature] Support ingesting nested array from parquet

xiaoyong-z opened the pull request:
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #16403

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
broker load currently not support ingesting nested array from parquet, this pr add support to it.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
- [x] 2.5 
- [x] 2.4
- [ ] 2.3
- [ ] 2.2

chaoyli commented on be/src/exec/parquet_scanner.cpp:
```diff
@@ -259,23 +272,17 @@ Status ParquetScanner::convert_array_to_column(ConvertFunc conv_func, size_t num
+    auto* filter_data = (&_chunk_filter.front()) + _chunk_start_idx;
+    auto st = conv_func(array, _batch_start_idx, num_elements, data_column, _chunk_start_idx, null_data, filter_data,
+                        &_conv_ctx);
+    if (st.ok()) {
```
```suggestion
    if (st.ok() && column->is_nullable())
```

chaoyli commented on be/src/exec/parquet_scanner.cpp:
```diff
@@ -244,6 +244,19 @@ Status ParquetScanner::new_column(const arrow::DataType* arrow_type, const SlotD
+        // only base types are supported, nested types are not supported
+        size_t depth_limit = type_desc->get_array_depth_limit();
+        auto arrow_list_check_depth_func = get_arrow_list_check_depth();
+        Status st = arrow_list_check_depth_func(array, depth_limit);
```
```suggestion
        RETURN_IF_ERROR(arrow_list_check_depth_func(array, depth_limit));
```

chaoyli commented on be/test/exec/test_data/parquet_data/generate_nest_array.py:
```diff
@@ -0,0 +1,41 @@
+
+output = "./nested_array.parquet"
+
+data = [
```
This file is not need.

xiaoyong-z committed:
[Feature] Support ingesting nested array from parquet

broker load currently not support ingesting nested array from parquet, this pr add support to it.

Signed-off-by: xyz <a997647204@gmail.com>

satanson commented on be/src/exec/arrow_to_starrocks_converter.cpp:
```diff
@@ -779,181 +779,192 @@ struct ArrowListConverter {
+        using ArrowArrayType = typename arrow::TypeTraits<T>::ArrayType;
+        using OffsetsType = typename T::offset_type;
+        if (child_array_start_idx && child_array_num_elements) {
+            auto child_array = down_cast<const ArrowArrayType*>(array);
```
can nested list  be LargeList or FixedSizeList?

xiaoyong-z replied:
add FixedSizeList in new pr

satanson commented on be/src/exec/arrow_to_starrocks_converter.cpp:
```diff
@@ -779,181 +779,192 @@ struct ArrowListConverter {
+    static arrow::Array* unnest_list_array(const arrow::Array* array) {
+        auto type_id = array->type_id();
+        if (is_list(type_id)) {
+            return get_list_array_child<arrow::ListType>(array, 0, 0);
```
2nd, 3rd paramaters are 0, looks weird.

wanpengfei-git commented:
run starrocks_admit_test

xiaoyong-z commented:
run starrocks_admit_test

xiaoyong-z commented:
@mergify rebase

xiaoyong-z commented:
run starrocks_admit_test
